### PR TITLE
En 7630/esdt sccall and invalid

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: process/smartContract/process.go [hunk 1]
```diff
@@ -221,129 +221,166 @@ func (sc *scProcessor) ExecuteSmartContractTransaction(
 	return returnCode, err
 }
 
-func (sc *scProcessor) doExecuteSmartContractTransaction(
```
Comment (iulianpascalau) [inline process/smartContract/process.go:224]: :+1: for splitting the large functions in smaller ones

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: go.mod [hunk 2]
```diff
@@ -46,3 +46,5 @@ require (
 )
 
 replace github.com/gogo/protobuf => github.com/ElrondNetwork/protobuf v1.3.2
+
+replace github.com/ElrondNetwork/arwen-wasm-vm => ../arwen-wasm-vm
```
Comment (ccorcoveanu) [inline go.mod:50]: remove local version

File: process/coordinator/transactionType.go [hunk 3]
```diff
@@ -75,11 +75,15 @@ func (tth *txTypeHandler) ComputeTransactionType(tx data.TransactionHandler) pro
 		return process.MoveBalance
 	}
 
+	funcName := tth.getFunctionFromArguments(tx.GetData())
+	isBuiltInFunction := tth.isBuiltInFunctionCall(funcName)
 	if isAsynchronousCallBack(tx) {
+		if isBuiltInFunction {
```
Comment (ccorcoveanu) [inline process/coordinator/transactionType.go:81]: From what I see we can get away with `if isBuiltInFunction {...` outside of this if and remove it from line 100. There is no case where `isBuiltInFunction == true` and we return something else. 
At line 96 we can just have `if !isDestInSelfShard`

**Review by ccorcoveanu — CHANGES_REQUESTED** [review]

File: process/smartContract/process.go [hunk 4]
```diff
@@ -221,129 +221,166 @@ func (sc *scProcessor) ExecuteSmartContractTransaction(
...
 		log.Debug("reloadLocalAccount error", "error", err.Error())
-		return 0, err
+		return nil, err
 	}
 
-	var results []data.TransactionHandler
-	results, err = sc.processVMOutput(vmOutput, txHash, tx, acntSnd, vmInput.CallType, vmInput.GasProvided)
-	if err != nil {
-		log.Trace("process vm output returned with problem ", "err", err.Error())
-		return vmOutput.ReturnCode, sc.ProcessIfError(acntSnd, txHash, tx, err.Error(), []byte(vmOutput.ReturnMessage), snapshot)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:312]: I do not think so. Will run importDB to fix whatever broken

File: process/coordinator/transactionType.go (hunk 3 above)
Comment (sasurobert) [reply process/coordinator/transactionType.go:81]: done

File: go.mod (hunk 2 above)
Comment (sasurobert) [reply go.mod:50]: will remove once arwen is in master

File: process/smartContract/process.go [hunk 5]
```diff
@@ -482,90 +523,256 @@ func (sc *scProcessor) resolveFailedTransaction(
...
+	acntSnd state.UserAccountHandler,
+	acntDst state.UserAccountHandler,
+	snapshot int,
+) (bool, *vmcommon.VMOutput, error) {
+	isSCCall, newVMInput, err := sc.isSCExecutionAfterBuiltInFunc(tx, vmInput, vmOutput, acntDst)
+	if !isSCCall {
+		return false, vmOutput, nil
+	}
+
+	userErrorVmOutput := &vmcommon.VMOutput{ReturnCode: vmcommon.UserError}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:681]: done

File: process/smartContract/process.go [hunk 6]
```diff
@@ -221,129 +221,166 @@ func (sc *scProcessor) ExecuteSmartContractTransaction(
...
+	acntSnd, acntDst state.UserAccountHandler,
+) (*vmcommon.VMOutput, error) {
 	if check.IfNil(acntDst) {
-		return 0, process.ErrNilSCDestAccount
+		return nil, process.ErrNilSCDestAccount
 	}
 
-	var vmExec vmcommon.VMExecutionHandler
-	vmExec, err = findVMByTransaction(sc.vmContainer, tx)
+	userErrorVmOutput := &vmcommon.VMOutput{ReturnCode: vmcommon.UserError}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:308]: done

File: go.mod (hunk 2 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply go.mod:50]: will do once arwen is in master

File: process/smartContract/process.go [hunk 7]
```diff
@@ -221,129 +221,166 @@ func (sc *scProcessor) ExecuteSmartContractTransaction(
 	return returnCode, err
 }
 
-func (sc *scProcessor) doExecuteSmartContractTransaction(
+func (sc *scProcessor) prepareExecution(
 	tx data.TransactionHandler,
 	acntSnd, acntDst state.UserAccountHandler,
-) (vmcommon.ReturnCode, error) {
+	builtInFuncCall bool,
+) (vmcommon.ReturnCode, *vmcommon.ContractCallInput, []byte, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:228]: there should be no more changes here.

**Review by iulianpascalau — APPROVED** [review]
Please add a description for this PR

File: go.mod (hunk 2 above)
Comment (sasurobert) [reply go.mod]: done

**Review by ccorcoveanu — APPROVED** [review]

**Pull request merged**